presto-native-execution/presto_cpp/main/LinuxMemoryChecker.cpp

```diff
@@ -0,0 +1,180 @@
+#include <folly/String.h>
+#include <folly/gen/Base.h>
+#include <folly/gen/File.h>
+#include <folly/gen/String.h>
+#include "presto_cpp/main/PeriodicMemoryChecker.h"
+#include "presto_cpp/main/common/Configs.h"
+
+#include <sys/stat.h>
+
+namespace facebook::presto {
+
+class LinuxMemoryChecker : public PeriodicMemoryChecker {
+ public:
+  explicit LinuxMemoryChecker(const PeriodicMemoryChecker::Config& config)
+      : PeriodicMemoryChecker(config) {
+    // Find out what cgroup version (v1 or v2) we have based on the directory
+    // it's mounted.
+    struct stat buffer;
+    if ((stat(kCgroupV1Path, &buffer) == 0)) {
+      statFile_ = kCgroupV1Path;
+    } else if ((stat(kCgroupV2Path, &buffer) == 0)) {
+      statFile_ = kCgroupV2Path;
+    } else {
+      statFile_ = "None";
+    }
+    LOG(INFO) << fmt::format("Using memory stat file {}", statFile_);
+  }
+
+  ~LinuxMemoryChecker() override {}
+
+  int64_t getUsedMemory() {
+    return systemUsedMemoryBytes();
+  }
+
+  void setStatFile(std::string statFile) {
+    statFile_ = statFile;
+    LOG(INFO) << fmt::format("Changed to using memory stat file {}", statFile_);
+  }
+
+ protected:
+  // Current memory calculation used is inactive_anon + active_anon
+  // Our first attempt was using memInfo memTotal - memAvailable.
+  // However memInfo is not containerized so we reserve this as a
+  // last resort.
+  //
+  // Next we tried to use what docker/kubernetes uses for their
+  // calculation. cgroup usage_in_bytes - total_inactive_files.
+  // However we found out that usage_in_bytes is a fuzz value
+  // and has a chance for the sync to occur after the shrink
+  // polling interval. This would result in double shrinks.
+  //
+  // Therefore we decided on values from the memory.stat file
+  // that are real time statistics. At first we tried to use
+  // the calculation suggested by the kernel team RSS+CACHE(+SWAP)
+  // However we noticed that this value was not closely related to the
+  // value in usage_in_bytes which is used to OOMKill. We then looked
+  // at all of the values in the stat file and decided that
+  // inactive_anon + active_anon moves closest to that of
+  // usage_in_bytes
+  //
+  // NOTE: We do not know if cgroup V2 memory.current is a fuzz
+  // value. It may be better than what we currently use. For
+  // consistency we will match cgroup V1 and change if
+  // necessary.
+  int64_t systemUsedMemoryBytes() override {
+    size_t memAvailable = 0;
+    size_t memTotal = 0;
+    size_t inactiveAnon = 0;
+    size_t activeAnon = 0;
+
+    if (statFile_ != "None") {
+      folly::gen::byLine(statFile_.c_str()) |
+          [&](const folly::StringPiece& line) -> void {
+        if (inactiveAnon == 0) {
+          inactiveAnon =
+              extractNumericConfigValueWithRegex(line, kInactiveAnonRegex);
+        }
+
+        if (activeAnon == 0) {
+          activeAnon =
+              extractNumericConfigValueWithRegex(line, kActiveAnonRegex);
+        }
```

Comment:
Can we return here `if (activeAnon != 0 && inactiveAnon != 0) { return; }`

Reply:
I'll put that as the first check